### ISSUE TITLE
AI-007 — DL Image Embeddings: Transfer Learning for Properties

### ISSUE THREAD

**Issue by Sebastian906** [Sebastian906/Nestorria-FullStack#154]
**Theme:** Deep Learning
**Priority:** P3
**Type:** Experiment

### Objective

Implement a property image embedding extraction pipeline using ResNet50 transfer learning, to improve recommendations and enable visual search.

### Context

Properties have images stored in Cloudinary. Image embeddings can improve recommendations (visually similar properties) and enable features like "search properties similar to this image". With 10 properties and ~50 images, the volume is experimental.

### Scope

- Image downloading from Cloudinary
- Preprocessing for ResNet50 (resize, normalize)
- Embedding extraction with pre-trained ResNet50 on ImageNet
- Embedding storage in PostgreSQL (pgvector)
- Visual similarity search endpoint
- "Visual similar" endpoint for a property

### Out of Scope

- ResNet50 fine-tuning
- Custom model training
- Property classification CNN
- Two-tower model (evaluate later)
- Real-time visual search

### Files to Create

```
ai-service/app/dl/
├── __init__.py
├── image/
│   ├── __init__.py
│   ├── preprocessor.py    # Image preprocessing
│   ├── embeddings.py      # ResNet50 embedding extractor
│   └── similarity.py      # Visual similarity search
ai-service/app/routers/
├── visual.py              # FastAPI router for visual search
ai-service/tests/dl/
├── test_image_preprocessor.py
├── test_image_embeddings.py
└── test_visual_similarity.py
```

### Files to Modify

- `ai-service/app/main.py` — register visual router
- `ai-service/requirements.txt` — torch, torchvision, Pillow
- `server/src/main/resources/application.properties` — pgvector extension

### Dependencies

- AI-001 (service scaffold)
- AI-002 (ML foundation)
- PostgreSQL with pgvector (configure in DB)

### Expected Implementation

1. **Preprocessor:**
   ```python
   class ImagePreprocessor:
       def __init__(self, size=(224, 224)):
           self.transform = transforms.Compose([
               transforms.Resize(size),
               transforms.ToTensor(),
               transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                  std=[0.229, 0.224, 0.225])
           ])
       
       def preprocess(self, image: Image.Image) -> torch.Tensor:
           return self.transform(image).unsqueeze(0)
   ```

2. **Embedding Extractor:**
   ```python
   class ImageEmbeddingExtractor:
       def __init__(self):
           self.model = models.resnet50(pretrained=True)
           self.model = nn.Sequential(*list(self.model.children())[:-1])
           self.model.eval()
       
       def extract(self, image: torch.Tensor) -> np.ndarray:
           with torch.no_grad():
               embedding = self.model(image)
           return embedding.flatten()  # 2048-dimensional vector
   ```

3. **Similarity Search:**
   ```python
   class VisualSimilarityEngine:
       def __init__(self):
           self.embeddings = {}  # property_id -> embedding
       
       def index_property(self, property_id: str, image_urls: List[str]):
           # Download, preprocess, extract, store
       
       def find_similar(self, property_id: str, top_k: int = 5) -> List[str]:
           # Cosine similarity against all indexed embeddings
       
       def search_by_image(self, image: Image.Image, top_k: int = 5) -> List[str]:
           # Search with uploaded image
   ```

4. **Storage:**
   - `property_image_embeddings` table with pgvector
   - IVFFlat index for efficient search
   - Metadata: property_id, image_url, embedding (vector(2048))

### API / Contract

```
GET /dl/visual/similar/{propertyId}?limit=5
Response:
{
  "propertyId": "...",
  "similar": [
    { "propertyId": "...", "similarity": 0.89 },
    { "propertyId": "...", "similarity": 0.85 }
  ],
  "model": "resnet50"
}

POST /dl/visual/search
Request (multipart):
  - image: File

Response:
{
  "results": [
    { "propertyId": "...", "similarity": 0.92 },
    { "propertyId": "...", "similarity": 0.88 }
  ]
}
```

### Data

- **Images:** ~50 images from 10 properties (Cloudinary)
- **Embeddings:** 2048 dimensions per image
- **Storage:** pgvector in PostgreSQL
- **Index:** IVFFlat with 10 lists (for 50 vectors)

### Security

- Endpoint requires authentication
- Rate limiting on visual search
- Do not expose Cloudinary URLs directly

### Testing

- Test preprocessor resizes and normalizes correctly
- Embeddings have dimension 2048
- Similarity scores between 0 and 1
- Search returns ordered results
- Indexing works correctly

### Observability

- Embedding extraction timing logs
- Index size logs
- Search latency
- CPU/GPU memory usage

### Acceptance Criteria

- [ ] `ImagePreprocessor.preprocess()` returns tensor (1, 3, 224, 224)
- [ ] `ImageEmbeddingExtractor.extract()` returns 2048-dim vector
- [ ] `VisualSimilarityEngine.index_property()` stores embedding
- [ ] `GET /dl/visual/similar/{id}` returns similar properties
- [ ] Cosine similarity in range [0, 1]
- [ ] Search works without GPU (CPU mode)
- [ ] pgvector extension installed and functional
- [ ] ResNet50 model loads correctly

### Risks

- **High:** Torch may be heavy (~2GB) → use torch CPU-only
- **Medium:** 50 images insufficient for evaluation → development only
- **Medium:** pgvector may not be installed → verify in docker-compose
- **Low:** ResNet50 may be slow on CPU → batch processing

### Notes

- **STATUS: EXPERIMENTAL** — Do not activate in production
- torch-cpu instead of torch-gpu to save space
- ResNet50 is a baseline — can be changed to EfficientNet later
- pgvector is included with PostgreSQL 16 (extension `vector`)
- When 500+ properties exist, evaluate search performance
